elasticsearch/client.go

```diff
@@ -18,133 +18,89 @@
+func (v8 clientV8) Search(index string, body io.Reader) (int, io.ReadCloser, error) {
+	response, err := v8.c.Search(
+		v8.c.Search.WithContext(context.Background()),
+		v8.c.Search.WithIndex(index),
+		v8.c.Search.WithBody(body),
+		v8.c.Search.WithTrackTotalHits(true),
+		v8.c.Search.WithPretty(),
+	)
 	if err != nil {
-		return nil, err
+		return 0, nil, err
 	}
-
-	return goelasticsearch.NewClient(goelasticsearch.Config{
-		Username:  config.Username,
-		Password:  config.Password,
-		Addresses: addresses,
-		Transport: &http.Transport{
-			Proxy:   proxy,
-			Dial:    dialer.Dial,
-			DialTLS: tlsDialer.Dial,
-		},
-	})
+	return response.StatusCode, response.Body, nil
 }
 
-// Validate ensures Hosts instance has at least one entry
-func (h Hosts) Validate() error {
-	if len(h) == 0 {
-		return errInvalidHosts
-	}
-	return nil
+type clientV7 struct {
+	c *v7.Client
 }
 
-func httpProxyURL(cfg *Config) (func(*http.Request) (*url.URL, error), error) {
-	if cfg.ProxyDisable {
-		return nil, nil
+// Search satisfies the Client interface for version 7
+func (v7 clientV7) Search(index string, body io.Reader) (int, io.ReadCloser, error) {
+	response, err := v7.c.Search(
+		v7.c.Search.WithContext(context.Background()),
+		v7.c.Search.WithIndex(index),
+		v7.c.Search.WithBody(body),
+		v7.c.Search.WithTrackTotalHits(true),
+		v7.c.Search.WithPretty(),
+	)
+	if err != nil {
+		return 0, nil, err
 	}
+	return response.StatusCode, response.Body, nil
+}
 
-	if cfg.ProxyURL == "" {
-		return http.ProxyFromEnvironment, nil
+// NewClient parses the given config and returns  a version-aware client as an interface
+func NewClient(config *Config) (Client, error) {
+	if config == nil {
+		return nil, errConfigMissing
 	}
 
-	proxyStr := cfg.ProxyURL
-	if !strings.HasPrefix(proxyStr, prefixHTTP) {
-		proxyStr = prefixHTTPSchema + proxyStr
-	}
-	u, err := url.Parse(proxyStr)
+	conn, err := connectionConfig(config)
 	if err != nil {
 		return nil, err
 	}
-	return http.ProxyURL(u), nil
-}
 
-func addresses(cfg *Config) ([]string, error) {
-	var addresses []string
-	for _, host := range cfg.Hosts {
-		address, err := common.MakeURL(cfg.Protocol, cfg.Path, host, defaultESPort)
-		if err != nil {
-			return nil, err
-		}
-		addresses = append(addresses, address)
+	t := &http.Transport{
+		Proxy:   conn.proxy,
+		Dial:    conn.dialer.Dial,
+		DialTLS: conn.tlsDialer.Dial,
 	}
-	return addresses, nil
+	return NewVersionedClient(config.Username, config.Password, conn.addresses, t)
 }
 
-func dialer(cfg *Config) (transport.Dialer, transport.Dialer, error) {
-	var tlsConfig *tlscommon.TLSConfig
-	var err error
-	if cfg.TLS.IsEnabled() {
-		if tlsConfig, err = tlscommon.LoadTLSConfig(cfg.TLS); err != nil {
-			return nil, nil, err
-		}
+// NewVersionedClient returns the right elasticsearch client for the current Stack version, as an interface
+func NewVersionedClient(user, pwd string, addresses []string, transport http.RoundTripper) (Client, error) {
```

Comment:
In my opinion the main driver for this should be compatibility of APM Server with Elasticsearch, for the latest minor of the last major and the first minor of the next major version. 
I don't see how this is achieved with this PR. The `version.GetDefaultVersion()` returns the `defaultBeatVersion` which will be the same for the running APM Server, independent of the Elasticsearch instances it is talking to.
In my opinion what would need to happen is that for every given ES address a request would be required to fetch the Elasticsearch version and then depending on that a client can be constructed. This would still open a couple of questions, like how to behave if the instances are running different versions, how to behave on upgrades, etc.

The second motivation for abstracting away the original elasticsearch client to keep backports to a minimum is not a strong motivation in my opinion. If the v7/v8 clients behave differently we will also need to implement both behaviours with this abstraction. I don't see a big difference effort wise, compared to changing the logic in the backport, except that the import would need to be updated for the version.

Reply:
Forgot to mention that the actual logic to address that compatibility issue is out the scope of this PR, since it is not so urgent.
This PR lays out the ground for it, I created a follow up ticket https://github.com/elastic/apm-server/issues/2975